flink-core/src/main/java/org/apache/flink/util/ExceptionUtils.java

```diff
@@ -314,6 +314,37 @@ else if (t instanceof Error) {
 		}
 	}
 
+	/**
+	 * Checks whether a throwable chain contains a specific type of exception and returns it. It deserializes
+	 * any {@link SerializedThrowable} that are found.
+	 *
+	 * @param throwable the throwable chain to check.
+	 * @param searchType the type of exception to search for in the chain.
+	 * @return Optional throwable of the requested type if available, otherwise empty
+	 */
+	public static <T extends Throwable> Optional<T> findSerializedThrowable(Throwable throwable, Class<T> searchType) {
+		if (throwable == null || searchType == null) {
+			return Optional.empty();
+		}
+
+		Throwable t = throwable;
+		while (t != null) {
+			if (searchType.isAssignableFrom(t.getClass())) {
+				return Optional.of(searchType.cast(t));
+			} else if (t.getClass().isAssignableFrom(SerializedThrowable.class)) {
+				Throwable next = ((SerializedThrowable) t).deserializeError(ClassLoader.getSystemClassLoader());
```

Comment:
I would pass in the class loader as a parameter because one might wanna use a user code class loader here.